xonsh/lexer.py

```diff
@@ -33,12 +32,27 @@
     GREATER,
     LESS,
     RIGHTSHIFT,
-    tokenize,
+    _tokenize,
     TokenError,
     HAS_WALRUS,
 )
 
 
+class LexToken(object):
+    __slots__ = ("type", "value", "lineno", "lexpos", "lexer")
+
+    def __str__(self):
+        return "LexToken(%s,%r,%d,%d)" % (
+            self.type,
+            self.value,
+            self.lineno,
+            self.lexpos,
+        )
+
+    def __repr__(self):
+        return str(self)
+
+
```

Comment:
I would prefer we use the `LexToken` from `ply`, if only to keep it obvious which of the parts are part of `ply` and which are custom to `xonsh`

Reply:
the idea is to not import lex. That will be saving some import cost.

Reply:
If there's a measurable difference then we can make the change.

Reply:
In my machine I get these numbers 

![Screenshot_20210909_202910](https://user-images.githubusercontent.com/6702219/132710809-5bef735c-764f-422f-a6f5-8f8d88b34424.png)

Reply:
I'm not sure that's a completely accurate representation of what the difference might be -- especially since we're importing other parts of `ply` elsewhere in the codebase, so those imports will be cached.
The difference should be measurable in startup time

Reply:
with the 
![image](https://user-images.githubusercontent.com/6702219/132805247-e32301a7-3e67-49a9-b8c9-3f67bf8063a2.png)

, it prints `Import cost ply.lex:  0.0027875900268554688`

Reply:
Hey @jnoortheen -- sorry I wasn't clearer -- I'm interested in `xonsh` general startup time, so running `time -p xonsh --no-rc -c "echo HI"` a number of times on `main` vs. this branch and determining if there's any noticeable improvement.

Reply:
That will not be same all the time. That is the actual cost.

Reply:
That is the cost assuming that we haven't imported any other part of `ply`, which I doubt.  

I don't trust benchmarks that exist in a vacuum.

Reply:
as you can see the numbers are small and doesn't make much difference. 
![xexec](https://user-images.githubusercontent.com/6702219/133125101-bd2d4566-f63a-4e7a-9972-131fe4240747.png)

importing `ply.yacc` will have minimum effect on importing `ply.lex` after. Since they are separate modules. The interpreter has to parse the file and built the tree right ?

Reply:
Based on those 4 measurements, it has no appreciable effect at all, so I'm not sure about the utility of breaking out one class from a vended subpackage

Reply:
OK understandable. I will revert this.